### PR TITLE
Remove alternate patterns to find fastq in demux flow cells

### LINKED ISSUE #3129: Is code for demultiplexing on sequencer still relevant?
## Description

There is a lot of infrastructure in cg for flow cells demultiplexed on the sequencer, which is currently unused because we don't demultiplex in the sequencer. It was tried in the past but it turned out to be too slow, and had some other issues that I don't remember clearly. 

Is this code still needed? Are we going to demultiplex in the sequencer at some point in the future?

## Suggested solution

Set for refinement for discussion if we should remove that code from cg.

## This can be closed when

- Decide what to do with this code
- implement the decision

## Blocked by

Refinement session

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
This PR introduces a very small change in the codebase, and a somehow bigger refactoring of the post-processing tests.

The change in the codebase closes #3129, removing the flow cell folder structure patterns involving a "Unaligned" directory, which is deprecated and leaving only two patterns: The flat structure, which has all the fastq files in the demux directory per flow cell, and the pattern for flow cells demultiplexed on the sequencer. This change is entirely in the file `cg/meta/demultiplex/utils.py`.

The refactoring of the tests fixes all the tests that the codebase change broke and adds testing of the remaining functionalities that is **coherent** with the sample sheet and demultiplex testing done before, i.e. using only the same _canonical_ flow cells.

### Added

- A fixture environment consisting of a store, a context and a Post-processing API based on the seven canonical flow cells for demultiplexing (2 HiSeqX, 2HiSeq2500, 2 NovaSeq6000 and 1 NovaSeqX) The new fixtures are:
  - `updated_store_with_demultiplexed_samples`
  - `updated_demultiplex_context`
  - `updated_demultiplex_context`
  - `seven_canonical_flow_cells`
  - `seven_canonical_flow_cells_selected_sample_ids`
- Files for the new demultiplexed flow cell fixtures
### Changed

- Removed all patterns involving `Unaligned` in `cg/meta/demultiplex/utils.py`
- Moved some fixtures to fixture pluggins

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b remove-bcl2fastq-06 -a
    ```

### How to test

- [x] `cg demultiplex finish flow-cell` and verify that the command works properly.
```shell
Post processing worked as expected
```

## Review

- [x] Tests executed by SD
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Deployed to stage:
```shell

```
- [ ] Deployed to production:
```shell

```

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -538,6 +552,16 @@ def demux_post_processing_api(
...
 
+@pytest.fixture
+def updated_demux_post_processing_api(
+    updated_demultiplex_context: CGConfig,
+    tmp_illumina_demultiplexed_flow_cells_directory,
+) -> DemuxPostProcessingAPI:
+    api = DemuxPostProcessingAPI(updated_demultiplex_context)
+    api.demultiplexed_runs_dir = tmp_illumina_demultiplexed_flow_cells_directory
+    return api
+
```
Comment (diitaz93) [inline tests/conftest.py:563]: same for this API

File: tests/conftest.py [hunk 2]
```diff
@@ -1254,6 +1278,26 @@ def store_with_demultiplexed_samples(
...
+        seven_canonical_flow_cells, seven_canonical_flow_cells_selected_sample_ids
+    ):
+        helpers.add_flow_cell_and_samples_with_sequencing_metrics(
+            flow_cell_name=flow_cell.id,
+            sequencer=flow_cell.sequencer_type,
+            sample_ids=sample_internal_ids,
+            store=store,
+        )
+    return store
+
```
Comment (diitaz93) [inline tests/conftest.py:1299]: and also replace the demultiplex store for this one

File: tests/fixture_plugins/demultiplex_fixtures/sample_fixtures.py [hunk 3]
```diff
@@ -79,13 +97,47 @@ def novaseq_6000_post_1_5_kits_bcl_convert_lims_samples(
...
     path = Path(novaseq_x_flow_cell_directory, "22F52TLT3_raw.json")
     return [FlowCellSampleBCLConvert.model_validate(sample) for sample in read_json(path)]
 
 
+@pytest.fixture
+def selected_novaseq_x_sample_ids() -> list[str]:
+    """Return a list of sample ids for a HiSeqX single index flow cell."""
+    return ["ACC13169A1", "ACC13170A6"]
+
+
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/sample_fixtures.py:118]: this fixture is needed to create the updated demultiplex store

File: tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py [hunk 4]
```diff
@@ -79,19 +79,39 @@ def novaseq_x_flow_cell(novaseq_x_flow_cell_dir: Path) -> FlowCellDirectoryData:
     return FlowCellDirectoryData(novaseq_x_flow_cell_dir)
 
 
+@pytest.fixture
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py:82]: this fixture is needed to create the updated demultiplex store

File: tests/meta/demultiplex/test_demux_post_processing.py [hunk 5]
```diff
@@ -40,107 +41,89 @@ class DemultiplexingScenario(BaseModel):
...
     """Test adding a demultiplexed flow cell to the databases with. Runs on each type of
     demultiplexing software and setting used."""
 
     # GIVEN a demultiplexed flow cell
-    flow_cell_demultiplexing_directory: str = request.getfixturevalue(scenario.flow_cell_directory)
-    flow_cell_name: str = request.getfixturevalue(scenario.flow_cell_name)
-    sample_internal_ids: list[str] = request.getfixturevalue(scenario.samples_ids)
+    flow_cell: FlowCellDirectoryData = request.getfixturevalue(flow_cell_fixture)
+    flow_cell_demultiplexing_directory: str = flow_cell.full_name
+    flow_cell_name: str = flow_cell.id
```
Comment (islean) [inline tests/meta/demultiplex/test_demux_post_processing.py:68]: Out of scope but I do not like the name vs id unclarities here.

File: tests/meta/demultiplex/test_demux_post_processing.py [hunk 6]
```diff
@@ -167,52 +150,63 @@ def test_get_all_demultiplexed_flow_cell_out_dirs(
...
+    selected_hiseq_x_single_index_sample_ids: list[str],
 ):
     # GIVEN a flow cell with undetermined fastqs in a non-pooled lane
 
+    # GIVEN that the flow cell has the sample sheet in housekeeper
+    add_and_include_sample_sheet_path_to_housekeeper(
+        flow_cell_directory=hiseq_x_single_index_flow_cell.path,
+        flow_cell_name=hiseq_x_single_index_flow_cell.id,
+        hk_api=updated_demux_post_processing_api.hk_api,
+    )
```
Comment (islean) [inline tests/meta/demultiplex/test_demux_post_processing.py:165]: was this step already added in the previous demux_post_processing_api? Or why has this behaviour changed?

**Review by islean — APPROVED** [review]
Looks good to me to me! Although I can surely have missed some fixtures, especially when the names have changed as well. 💯

File: tests/fixture_plugins/demultiplex_fixtures/name_fixtures.py [hunk 7]
```diff
@@ -47,6 +47,33 @@ def flow_cell_directory_name_demultiplexed_with_bcl_convert(
...
+@pytest.fixture(scope="session")
+def flow_cell_name_demultiplexed_with_bcl_convert() -> str:
+    return "HY7FFDRX2"
+
+
+@pytest.fixture(scope="session")
+def flow_cell_directory_name_demultiplexed_with_bcl_convert_flat(
+    flow_cell_name_demultiplexed_with_bcl_convert: str,
+):
+    """Return the name of a flow cell directory that has been demultiplexed with Bcl Convert using a flat output directory structure."""
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply tests/fixture_plugins/demultiplex_fixtures/name_fixtures.py:59]: You are right. I will add this to the next PR. There are still many fixture to fix

File: tests/meta/demultiplex/test_demux_post_processing.py (hunk 5 above)
Comment (diitaz93) [reply tests/meta/demultiplex/test_demux_post_processing.py:68]: This is an old discussion, see #2264, which I am totally in favour of bringing it back to life, but it is a big refactoring

File: tests/meta/demultiplex/test_demux_post_processing.py (hunk 6 above)
Comment (diitaz93) [reply tests/meta/demultiplex/test_demux_post_processing.py:165]: The fixture previously used (`low_cell_directory_name_demultiplexed_with_bcl_convert_flat`) added the sample sheet to HK in its definition. This one does not. I think it is debatable if the fixture of a file should include it to HK right away, but this is out of scope

**Pull request merged**